src/main/scala/com/fulcrumgenomics/bam/ConsensusCaller.scala

```diff
@@ -0,0 +1,369 @@
+      PhredValue(Math.min(SAMUtils.MAX_PHRED_SCORE, p.value)) // cap the quality
+    }
+    (consensusBases.mkString, consensusQualitiesScaled)
+  }
+
+  /** Adjusts the given base qualities.  The base qualities are first shifted by `baseQualityShift`, then capped using
+    * `maxBaseQuality`, and finally the `errorRatePostUmi` is incorporated.
+    */
+  private[bam] def adjustBaseQualities(quals: Seq[PhredValue],
+                                       maxBaseQuality: PhredValue   = DefaultMaxBaseQuality,
+                                       baseQualityShift: PhredValue = DefaultBaseQualityShift,
+                                       errorRatePostUmi: PhredValue = DefaultErrorRatePostUmi
+                                      ): Seq[PhredValue] = {
+    quals.map { qual =>
+      // shift the base qualities, then cap it
+      PhredValue(Math.max(0.0, qual.value - baseQualityShift.value)) match {
+        case q if maxBaseQuality < q => maxBaseQuality
+        case q => q
+      }
+    }.map { phred =>
+      // Pr(err) = Pr(no post-UMI error AND any sequencing error) + Pr(any post-UMI error and no sequencing error) +
+      //             Pr(post-UMI error AND sequencing error)
+      // The last term tries to capture the case where a post-UMI error modifies the base (ex. A->C) but a sequencing
+      // error calls it the correct base (C->A).  Only 2/3 times will the two errors result in the incorrect base.
+      probabilityOfErrorTwoTrials(errorRatePostUmi, phred)
+    }
+  }
+
+  /** Computes the probability of seeing an error in the base sequence if there are two independent error processes.
+    * We sum three terms:
+    * 1. the probability of an error in trial one and no error in trial two: Pr(A=Error, B=NoError).
+    * 2. the probability of no error in trial one and an error in trial two: Pr(A=NoError, B=Error).
+    * 3. the probability of an error in both trials, but when the second trial does not reverse the error in first one, which
+    *    for DNA (4 bases) would only occur 2/3 times: Pr(A=x->y, B=y->z) * Pr(x!=z | x!=y, y!=z, x,y,z \in {A,C,G,T})
+    */
+  private[bam] def probabilityOfErrorTwoTrials(prErrorTrialOne: PhredValue, prErrorTrialTwo: PhredValue): PhredValue = {
+    val pr1 = prErrorTrialOne       * prErrorTrialTwo.inv()
+    val pr2 = prErrorTrialOne.inv() * prErrorTrialTwo
+    val pr3 = prErrorTrialOne       * prErrorTrialTwo       * 2.0 / 3.0
+    pr1 + pr2 + pr3
+  }
+
+  /** The type of consensus read to output. */
+  private object ReadType extends Enumeration {
+    val Fragment, FirstOfPair, SecondOfPair = Value
+  }
+
+  /** Gets the longest common prefix of the given strings, None if there is only one string or if there is an empty string. */
+  @annotation.tailrec
+  private def longestCommonPrefix(strs: Iterable[String], accu: Option[String] = None): Option[String] = {
+    if (strs.exists(_.isEmpty) || strs.size <= 1) accu
+    else {
+      val firsts = strs.map(_.head)
```

Comment:
This could be more efficient by stopping at the first non-matching character.  E.g. 
```scala
val first = strs.head.head
if (strs.tail.forall(_.head == first)) recurse
else accu
```